### ISSUE TITLE
Multiband support for semantic segmentation

### ISSUE THREAD

**Issue by lewfish** [azavea/raster-vision#961]
This issue is for adding support for training models on > 3 bands for semantic segmentation. The main idea is to save and use images in numpy array format instead of PNG format. Adding support for other tasks, and adding automated testing and documentation is outside the scope of this issue.

## Testing Strategy
* First, make sure you can run the ISPRS Potsdam example since we will base our test off of it. See https://docs.rastervision.io/en/latest/examples.html#semantic-segmentation-isprs-potsdam and https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_backend/rastervision/pytorch_backend/examples/semantic_segmentation/isprs_potsdam.py
* Modify the example so that it has an extra `multiband` argument [here](https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_backend/rastervision/pytorch_backend/examples/semantic_segmentation/isprs_potsdam.py#L15)
* If the `multiband` option is True, then use all four bands as the channel order [here](https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_backend/rastervision/pytorch_backend/examples/semantic_segmentation/isprs_potsdam.py#L59).
* When developing this, you will want to run it locally, and using `-a test True`. I also find it very helpful to use the debugger. I use the debugger in Visual Studio Code, and have some support for this using `docker/run --debug` and inside the container `./scripts/debug rastervision.pipeline.cli run ...`. You will need to configure the remote debugger in VS Code using something like the following to do volume mounting:
```
        {
            "name": "Raster Vision: Remote Attach",
            "type": "python",
            "request": "attach",
            "port": 3003,
            "host": "localhost",
            "pathMappings": [
                {
                    "localRoot": "/Users/lfishgold/projects/raster-vision",
                    "remoteRoot": "/opt/src"
                }
            ],
            "justMyCode": false
        },
```
* Publish image to ECR using `docker/ecr_publish`, and run the experiment on Batch.
* Check that ndarray files are actually produced in the `chip` directory on S3. The scores in the `eval` directory should be better than or equal to running with only three bands.

## Implementation Strategy
* In `rastervision.pytorch_backend`, write chips in [`ndarray`](https://numpy.org/doc/stable/reference/generated/numpy.save.html) format in https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_backend/rastervision/pytorch_backend/pytorch_semantic_segmentation.py#L15 We should use ndarrays of a type that matches the resolution of the GeoTIFF in order to save space. For example, if the GeoTIFF is uint16, then we should use an ndarray with `dtype` equal to `np.uint16`.
* In `rastervision.pytorch_learner`,  make dataset read chips in ndarray format. We should leave the existing ability to read PNG files intact. (This is because the `rastervision.pytorch_learner` package is meant to be usable outside of a geospatial context, and PNGs are a more standard way of distributing training data.) We can do this by adding a new data format in https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/semantic_segmentation_learner_config.py#L10 and add a new `Dataset` that handles `ndarray` datasets that is similar to https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/semantic_segmentation_learner.py#L23. Alternatively, we could just modify the `SemanticSegmentationDataset` to handle both png and ndarray files. That could be simpler, but I'm not sure.
  * A potential issue is that `albumentations` (the data augmentation library) might not be able to handle > 3 bands (I haven't looked into it).
  * Another issues is that dividing by 255.0 to get a number between 0 and 1 which the model expects (https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/semantic_segmentation_learner.py#L42) will no longer be correct if the ndarrays have more resolution than uint8.
*  To build a model that can handle > 3 bands, we will need to modify the DeeplabV3 model before it is returned in the `build_model` method: https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/semantic_segmentation_learner.py#L52 @jamesmcclain has code for doing this. Handling the `pretrained` option correctly will require some work because the pretrained model only natively handles 3 bands.
- The `plot_xyz` method needs to be modified so it picks three of the bands for display purposes: https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/semantic_segmentation_learner.py#L132 To select the bands to use, a new option can be added to `SemanticSegmentationDataConfig` https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/semantic_segmentation_learner_config.py#L15
* Note that any configuration options added to `rastervision.pytorch_learner` will need to also be added to https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_backend/rastervision/pytorch_backend/pytorch_semantic_segmentation_config.py#L14 and then forwarded similar to https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_backend/rastervision/pytorch_backend/pytorch_semantic_segmentation_config.py#L21

**Comment by lewfish** [comment]
@jamesmcclain Could you please link to the code you have for modifying a Deeplab model to handle > 3 bands (including any needed to use Imagenet pretrained models)?

**Comment by jamesmcclain** [comment]
> @jamesmcclain Could you please link to the code you have for modifying a Deeplab model to handle > 3 bands (including any needed to use Imagenet pretrained models)?

I am not sure if I have any materials that pertains specifically to ImageNet, but [this code](https://github.com/geotrellis/deeplab-nlcd/blob/73c5afda772b36ef4ab705660aaae040a8508ca9/python/architectures/deeplab-resnet101-stock.py#L35-L37) shows how to modify an architecture (in this case DeepLab with a ResNet-101 backbone) to use more than three channels.

**Comment by AdeelH** [comment]
>A potential issue is that albumentations (the data augmentation library) might not be able to handle > 3 bands (I haven't looked into it).

It does seem to support it: 
https://github.com/albumentations-team/albumentations/commit/c028a9557cc960da11720a0a505a19cdd4fe0b24
https://github.com/albumentations-team/albumentations/pull/638/files

>Another issues is that dividing by 255.0 to get a number between 0 and 1 which the model expects (https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/semantic_segmentation_learner.py#L42) will no longer be correct if the ndarrays have more resolution than uint8.

Maybe we could add a `per_channel_normalization_const` option of type `Union[float, Iterable[float]]` that is 255. by default but can be changed to something like `(255., 255., 255. 65535.)` etc.

>The plot_xyz method needs to be modified so it picks three of the bands for display purposes: https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/semantic_segmentation_learner.py#L132 To select the bands to use, a new option can be added to SemanticSegmentationDataConfig https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/semantic_segmentation_learner_config.py#L15

This might be a bit too elaborate, but one way could be to specify "channel groups" such that `channel_display_groups = [(0, 1, 2), (3,)]` for an RGBElev input would result in two side-by-side images like so: 

![](https://2rct3i2488gxf9jvb1lqhek9-wpengine.netdna-ssl.com/wp-content/uploads/2019/08/potsdam-data-example.png)

**Comment by lewfish** [comment]
To recap what was discussed yesterday:

> > Another issues is that dividing by 255.0 to get a number between 0 and 1 which the model expects (https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/semantic_segmentation_learner.py#L42) will no longer be correct if the ndarrays have more resolution than uint8.
> 
> Maybe we could add a `per_channel_normalization_const` option of type `Union[float, Iterable[float]]` that is 255. by default but can be changed to something like `(255., 255., 255. 65535.)` etc.

I would just use the `RasterSource.get_dtype()` method and only add the `per_channel_normalization_const` if we really need it for some use case.

> 
> > The plot_xyz method needs to be modified so it picks three of the bands for display purposes: https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/semantic_segmentation_learner.py#L132 To select the bands to use, a new option can be added to SemanticSegmentationDataConfig https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/semantic_segmentation_learner_config.py#L15
> 
> This might be a bit too elaborate, but one way could be to specify "channel groups" such that `channel_display_groups = [(0, 1, 2), (3,)]` for an RGBElev input would result in two side-by-side images like so

This sounds good. In addition, having the ground truth labels in a separate plot with a legend regardless of using multiband would be nice.

**Comment by AdeelH** [comment]
> I would just use the `RasterSource.get_dtype()`

`RasterSource` is only used in the chipping stage and we don't need to normalize then. I think it makes sense to just normalize based on the `dtype` of the array that the Dataset reads in. I've implemented it like this for now:
```python
def load_np_normalized(path: Union[IO, str, Path]):
    arr = np.load(path)
    dtype = arr.dtype
    if np.issubdtype(dtype, np.unsignedinteger):
        max_val = np.iinfo(dtype).max  # 255 for np.uint8
        arr = arr.astype(np.float32) / max_val
    return arr
```

> This sounds good. In addition, having the ground truth labels in a separate plot with a legend regardless of using multiband would be nice.

This is what I have now. Let me know if this works.

![example](https://user-images.githubusercontent.com/13014700/89795227-935e2800-db41-11ea-8418-eb69c7993396.png)

**Comment by lewfish** [comment]
Both of these look good.
